### PR TITLE
Feature/#454 open files/handles support

### PR DESCRIPTION
**Pull request by spyhunter99**

### CODE REVIEW

File: oshi-core/src/main/java/oshi/software/os/OSProcess.java [hunk 1]
```diff
@@ -523,4 +524,24 @@ public void setBytesRead(long bytesRead) {
...
+     * @param count 
+     */
+    public void setOpenFiles(long count){
+        this.handles = count;
+    }
+    
+    /**
+     * Sets the number of open file handles (or network connections) that 
+     * belongs to the process
+     * @since 3.5.0
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/OSProcess.java]: While this may be good documentation we've not used it at all in the past; also this may be a 3.4.x patch release....

File: oshi-core/src/main/java/oshi/software/os/linux/LinuxOperatingSystem.java [hunk 2]
```diff
@@ -262,6 +262,15 @@ public OSProcess getProcess(int pid) {
         proc.setBytesRead(ParseUtil.parseLongOrDefault(MapUtil.getOrDefault(io, "read_bytes", ""), 0L));
         proc.setBytesWritten(ParseUtil.parseLongOrDefault(MapUtil.getOrDefault(io, "write_bytes", ""), 0L));
         
+        //gets the open files count
+        try {
+            String openFilesString = ExecutingCommand.getFirstAnswer(String.format("ls -f /proc/%d/exe | wc -l", pid));
+            if (openFilesString!=null)
+                proc.setOpenFiles(Long.parseLong(openFilesString));
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/linux/LinuxOperatingSystem.java]: Please use the Util's `parseLongOrDefault()` method to catch any errors here (and elsewhere, won't repeat this comment).  I believe that will avoid having to catch exceptions at this level.

File: oshi-core/src/main/java/oshi/software/os/linux/LinuxOperatingSystem.java [hunk 3]
```diff
@@ -262,6 +262,15 @@ public OSProcess getProcess(int pid) {
         proc.setBytesRead(ParseUtil.parseLongOrDefault(MapUtil.getOrDefault(io, "read_bytes", ""), 0L));
         proc.setBytesWritten(ParseUtil.parseLongOrDefault(MapUtil.getOrDefault(io, "write_bytes", ""), 0L));
         
+        //gets the open files count
+        try {
+            String openFilesString = ExecutingCommand.getFirstAnswer(String.format("ls -f /proc/%d/exe | wc -l", pid));
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/linux/LinuxOperatingSystem.java]: Shouldn't this be `/proc/%d/fd`?  See [proc man page](http://man7.org/linux/man-pages/man5/proc.5.html).

File: oshi-core/src/main/java/oshi/software/os/unix/freebsd/FreeBsdOperatingSystem.java [hunk 4]
```diff
@@ -143,9 +145,17 @@ public OSProcess getProcess(int pid) {
...
-        }
+            //gets the open files count
+            try {
+                String openFilesString = ExecutingCommand.getFirstAnswer(String.format("lsof -p %d | wc -l", pid));
+                if (openFilesString!=null)
+                    fproc.setOpenFiles(Long.parseLong(openFilesString));
+            } catch (Exception ex) {
+                LOG.warn("Couldn't find get open file count for pid {}: {}", pid, ex);
+            }
+                // 'top -bm io' gives read/write counts, not bytes
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/unix/freebsd/FreeBsdOperatingSystem.java]: this comment looks out of place, should it be located elsewhere?

File: oshi-core/src/main/java/oshi/software/os/windows/WindowsOperatingSystem.java [hunk 5]
```diff
@@ -56,11 +56,11 @@
     // For WMI Process queries
     private static String processProperties = "Name,ExecutablePath,CommandLine,ExecutionState,ProcessID,ParentProcessId"
             + ",ThreadCount,Priority,VirtualSize,WorkingSetSize,KernelModeTime,UserModeTime,CreationDate"
-            + ",ReadTransferCount,WriteTransferCount,__PATH,__PATH";
+            + ",ReadTransferCount,WriteTransferCount,__PATH,__PATH,HandleCount";
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/windows/WindowsOperatingSystem.java]: purely preference, I'd prefer to move the HandleCount before the two process queries, to leave them at the end...

File: oshi-core/src/main/java/oshi/software/os/OSProcess.java (hunk 1 above)
Comment (spyhunter99) [reply oshi-core/src/main/java/oshi/software/os/OSProcess.java]: ok, what should this return if it's not supported or could not be acquired? -1? 0?

File: oshi-core/src/main/java/oshi/software/os/windows/WindowsOperatingSystem.java (hunk 5 above)
Comment (spyhunter99) [reply oshi-core/src/main/java/oshi/software/os/windows/WindowsOperatingSystem.java]: so after `WriteTransferCount`?

Comment (dbwiddis) [reply oshi-core/src/main/java/oshi/software/os/windows/WindowsOperatingSystem.java]: Yep... again, purely preference to make it easier to find those processes at the end.

File: oshi-core/src/main/java/oshi/software/os/OSProcess.java (hunk 1 above)
Comment (dbwiddis) [reply oshi-core/src/main/java/oshi/software/os/OSProcess.java]: I would suggest either deleting the @since, or updating all the rest of the code to add the @since everywhere it's needed (which is pretty much everywhere).

File: oshi-core/src/main/java/oshi/software/os/OSProcess.java [hunk 6]
```diff
@@ -523,4 +524,22 @@ public void setBytesRead(long bytesRead) {
...
+     * @param count 
+     */
+    public void setOpenFiles(long count){
+        this.handles = count;
+    }
+    
+    /**
+     * Sets the number of open file handles (or network connections) that 
+     * belongs to the process
+     * @return open files or -1 if
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/OSProcess.java]: -1 if... ? :)

File: oshi-core/src/main/java/oshi/software/os/OSProcess.java [hunk 7]
```diff
@@ -53,6 +53,7 @@
     private long upTime;
     private long bytesRead;
     private long bytesWritten;
+    private long handles;
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/OSProcess.java]: I think I'd prefer the internal variable name match the getter name.

File: oshi-core/src/main/java/oshi/software/os/mac/MacOperatingSystem.java [hunk 8]
```diff
@@ -195,6 +196,14 @@ public OSProcess getProcess(int pid) {
         proc.setBytesRead(bytesRead);
         proc.setBytesWritten(bytesWritten);
         proc.setCommandLine(getCommandLine(pid));
+        //gets the open files count
+        try {
+            String openFilesString = ExecutingCommand.getFirstAnswer(String.format("lsof -p %d | wc -l", pid));
+            if (openFilesString!=null)
+                proc.setOpenFiles(Long.parseLong(openFilesString));
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/mac/MacOperatingSystem.java]: another `ParseUtil.parseLongOrDefault()` needed here.

File: oshi-core/src/test/java/oshi/software/os/OperatingSystemTest.java [hunk 9]
```diff
@@ -73,6 +73,7 @@ public void testOperatingSystem() {
         assertTrue(proc.getStartTime() >= 0);
         assertTrue(proc.getBytesRead() >= 0);
         assertTrue(proc.getBytesWritten() >= 0);
+        assertTrue(proc.getOpenFiles()>= -1);
```
Comment (dbwiddis) [inline oshi-core/src/test/java/oshi/software/os/OperatingSystemTest.java]: my code formatter will add a space here.  Wouldn't hold up the commit for this but if you're making other changes, may as well prevent the space being added in the next unrelated commit :)

**Comment by dbwiddis** [comment]
Also don't forget your changelog entry!

**Comment by dbwiddis** [comment]
Annnnd.... did you add the code to retrieve the stat in the JSON version?

**Comment by spyhunter99** [comment]
why is there this much code duplication? probably simpler to have a json utils class or just use jackson's data bind to convert the objects to/from json

**Comment by dbwiddis** [comment]
Yes, we know.  [Sometime in the unspecified future we'd like this fixed](https://github.com/oshi/oshi/issues/365#issuecomment-302506211).  Until then, you're stuck with what I (a non-SDE) hacked together. :)
